kusd/validator/states.go

```diff
@@ -91,6 +52,60 @@ func (val *Validator) notLoggedInState() stateFn {
 	return val.newElectionState
 }
 
+func (val *Validator) nodeNeedsToDeposit() bool {
+	// @NOTE (rgeraldes) - sync was already done at this point and by default the investors will be
+	// part of the initial set of validators - no need to make a deposit if the block number is 0
+	// since these validators will be marked as voters from the start
+	isGenesis, err := val.network.IsGenesisVoter(&bind.CallOpts{}, val.account.Address)
+	if err != nil {
+		log.Crit("Failed to verify the voter information", "err", err)
+	}
+	return !isGenesis || val.isGenesisRejoining(isGenesis)
+}
+
+func (val *Validator) isGenesisRejoining(isGenesis bool) bool {
+	return isGenesis && val.isNotFirstBlock()
+}
+
+func (val *Validator) isFirstBlock() bool {
+	return val.chain.CurrentBlock().NumberU64() == 0
+}
+
+func (val *Validator) isNotFirstBlock() bool {
+	return !val.isFirstBlock()
+}
+
+func (val *Validator) waitToParticipate() {
+	// Subscribe events from blockchain
+	chainHeadCh := make(chan core.ChainHeadEvent)
+	chainHeadSub := val.chain.SubscribeChainHeadEvent(chainHeadCh)
+	defer chainHeadSub.Unsubscribe()
+
+	log.Info("Waiting confirmation to participate in the consensus")
+L:
+	for {
+		select {
+		case _, ok := <-chainHeadCh:
+			if !ok {
+				log.Info("Failed reading on Header channel")
+			}
+
+			isVoter := val.assertGetIsVoter()
+			if isVoter {
+				break L
+			}
+		}
+	}
+}
+
+func (val *Validator) assertGetIsVoter() bool {
```

Comment:
What's the difference between `assert...` and `is...`? both return a bool

Reply:
assert as in fail otherwise, there is a log.Crit that panic/exits so this func implies get or fail. It's a hack as I think most errors should be recoverable, or at least managed and bubbled up

Reply:
I see, didn't realise the `Crit` panics. Thanks!